cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/store/preview/DefaultPreviewStoreTest.java

```diff
@@ -122,4 +127,80 @@ public void testPreviewInfo() throws IOException {
     Assert.assertEquals(runId, store.getProgramRunId(applicationId));
     Assert.assertEquals(status, store.getPreviewStatus(applicationId));
   }
+
+  @Test
+  public void testPreviewWaitingRequests() {
+    String json = "{ \"id\": \"runner-1\"}";
+    JsonObject pollerInfo = new JsonParser().parse(json).getAsJsonObject();
+
+    Assert.assertEquals(0, store.getAllInWaitingState().size());
+
+    RunId id1 = RunIds.generate();
+    ApplicationId applicationId = new ApplicationId("ns1", id1.getId());
+    store.addToWaitingState(applicationId, getAppRequest());
+    List<PreviewRequest> allWaiting = store.getAllInWaitingState();
+    Assert.assertEquals(1, allWaiting.size());
+
+    AppRequest appRequest = allWaiting.get(0).getAppRequest();
+    Assert.assertNotNull(appRequest);
+    Assert.assertNotNull(appRequest.getPreview());
+    Assert.assertEquals("WordCount", appRequest.getPreview().getProgramName());
+    store.setPreviewRequestPollerInfo(applicationId, pollerInfo);
+
+    Assert.assertEquals(0, store.getAllInWaitingState().size());
+
+    // add 2 requests to the queue
+    ApplicationId applicationId2 = new ApplicationId("ns1", RunIds.generate().getId());
+    store.addToWaitingState(applicationId2, getAppRequest());
+    ApplicationId applicationId3 = new ApplicationId("ns1", RunIds.generate().getId());
+    store.addToWaitingState(applicationId3, getAppRequest());
+
+    allWaiting = store.getAllInWaitingState();
+    Assert.assertEquals(2, allWaiting.size());
+    Assert.assertEquals(applicationId2, allWaiting.get(0).getProgram().getParent());
+    Assert.assertEquals(applicationId3, allWaiting.get(1).getProgram().getParent());
+
+    store.setPreviewRequestPollerInfo(applicationId2, pollerInfo);
+    allWaiting = store.getAllInWaitingState();
+    Assert.assertEquals(1, allWaiting.size());
+    Assert.assertEquals(applicationId3, allWaiting.get(0).getProgram().getParent());
+
+    store.setPreviewRequestPollerInfo(applicationId3, pollerInfo);
+    allWaiting = store.getAllInWaitingState();
+    Assert.assertEquals(0, allWaiting.size());
+  }
+
+  private AppRequest getAppRequest() {
+    String appRequestWithSchedules = "{\n" +
```

Comment:
Why use json string to create an `AppRequest`?? It's better to new the object directly

Reply:
fixed.